src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp

```diff
@@ -0,0 +1,781 @@
+      }
+
+      // Set the specified bits in the polynomial
+      for(const auto i : set_bits) {
+         p[i] = 1;
+      }
+
+      index = end_index;
+   }
+
+   // Check that the remaining bit positions are all zero (strong unforgeability)
+   const auto remaining = bit_positions.take(bit_positions.remaining());
+   if(!std::all_of(remaining.begin(), remaining.end(), [](auto b) { return b == 0; })) {
+      return std::nullopt;
+   }
+
+   BOTAN_DEBUG_ASSERT(hint.ct_validate_value_range(0, 1));
+   return hint;
+}
+
+}  // namespace
+
+/**
+ * NIST FIPS 204 IPD, Algorithm 16 (pkEncode)
+ */
+DilithiumSerializedPublicKey dilithium_encode_public_key(StrongSpan<const DilithiumSeedRho> rho,
+                                                         const DilithiumPolyVec& t1,
+                                                         const DilithiumConstants& mode) {
+   DilithiumSerializedPublicKey pk(mode.public_key_bytes());
+   BufferStuffer stuffer(pk);
+
+   stuffer.append(rho);
+   for(const auto& p : t1) {
+      dilithium_poly_pack_t1(p, stuffer);
+   }
+
+   BOTAN_ASSERT_NOMSG(stuffer.full());
+   return pk;
+}
+
+/**
+ * NIST FIPS 204 IPD, Algorithm 17 (pkDecode)
+ */
+std::pair<DilithiumSeedRho, DilithiumPolyVec> dilithium_decode_public_key(
+   StrongSpan<const DilithiumSerializedPublicKey> pk, const DilithiumConstants& mode) {
+   if(pk.size() != mode.public_key_bytes()) {
+      throw Decoding_Error("Dilithium: Invalid public key length");
+   }
+
+   BufferSlicer slicer(pk);
+   auto rho = slicer.copy<DilithiumSeedRho>(DilithiumConstants::SEED_RHO_BYTES);
+
+   DilithiumPolyVec t1(mode.k());
+   for(auto& p : t1) {
+      dilithium_poly_unpack_t1(p, slicer);
+   }
+   BOTAN_ASSERT_NOMSG(slicer.empty());
+
+   return {std::move(rho), std::move(t1)};
+}
+
+/**
+ * NIST FIPS 204 IPD, Algorithm 18 (skEncode)
+ */
+DilithiumSerializedPrivateKey dilithium_encode_private_key(StrongSpan<const DilithiumSeedRho> rho,
+                                                           StrongSpan<const DilithiumHashedPublicKey> tr,
+                                                           StrongSpan<const DilithiumSigningSeedK> key,
+                                                           const DilithiumPolyVec& s1,
+                                                           const DilithiumPolyVec& s2,
+                                                           const DilithiumPolyVec& t0,
+                                                           const DilithiumConstants& mode) {
+   DilithiumSerializedPrivateKey sk(mode.private_key_bytes());
+   BufferStuffer stuffer(sk);
+
+   stuffer.append(rho);
+   stuffer.append(key);
+   stuffer.append(tr);
+
+   for(const auto& p : s1) {
+      dilithium_poly_pack_eta(p, stuffer, mode);
+   }
+
+   for(const auto& p : s2) {
+      dilithium_poly_pack_eta(p, stuffer, mode);
+   }
+
+   for(const auto& p : t0) {
+      dilithium_poly_pack_t0(p, stuffer);
+   }
+
+   BOTAN_ASSERT_NOMSG(stuffer.full());
```

Comment:
Those became ubiquitous and at this point I do start wondering whether we should just put that in the destructors of `Stuffer` and `Slicer` (perhaps explicitly disengagable). What do y'all think?

Reply:
Sounds good to me.